src/database-sqlite3.cpp

```diff
@@ -62,88 +68,484 @@ SQLite format specification:
+				<< m_journal_mode << "' on database. Mode is now: '"
+				<< new_mode << "'" << std::endl;
+		}
+		if (previous_mode != new_mode) {
+			infostream << "SQLite3: changed database journal mode from '"
+				<< previous_mode << "' to '" << new_mode << "'"
+				<< std::endl;
+		}
+		m_journal_mode = new_mode;
+	} else {
+		infostream << "SQLite3: database journal mode is '"
+			<< previous_mode << "' (not changed)" << std::endl;
+	}
+
+	if (m_journal_mode == "wal") {
+		SQLOK_ERRSTREAM(sqlite3_wal_autocheckpoint(m_database, 0),
+			"Failed to disable auto-checkpointing for WAL mode");
+	}
+}
+
+
+void SQLite3::setSynchronousLevel()
+{
+	if (m_synchronous == -1)
+		m_synchronous = (m_journal_mode == "wal") ? 1 : 2;
+	std::string query_str = std::string("PRAGMA synchronous = ")
+			 + itos(m_synchronous);
+	SQLOK(sqlite3_exec(m_database, query_str.c_str(), NULL, NULL, NULL),
+		"SQLite3: Failed to modify sqlite3 synchronous level");
+	infostream << "SQLite3: database synchronous level set to: "
+		<< m_synchronous << std::endl;
+}
+
+
+void SQLite3::startWALCheckpointThread()
+{
+	if (!m_database)
+		throw DatabaseException("SQLite3: Internal error: trying to start"
+			" WAL thread, but database is not yet open");
+
+	if (m_journal_mode != "wal" || getDBJournalMode() != "wal")
+		throw DatabaseException("SQLite3: Internal error: trying to start"
+			" WAL thread, but journal mode is not 'wal'");
+
+	if (m_walCheckpointThread)
+		throw DatabaseException("SQLite3: Internal error: trying to start"
+			" WAL thread, but it is already running");
+
+	setAutoCheckpoint(false);
+	m_walCheckpointThread = new SQLite3CheckpointThread(*this);
+	m_walCheckpointThread->start();
+}
+
+void SQLite3::stopWALCheckpointThread()
+{
+	if (m_walCheckpointThread) {
+		if (!m_walCheckpointThread->shutdown(10)) {
+			errorstream << "SQLite3: WAL thread fails to terminate - killing it"
+				<< std::endl;
+			m_walCheckpointThread->stop();
+		}
+		m_walCheckpointThread->wait();
+		delete m_walCheckpointThread;
+		m_walCheckpointThread = NULL;
+	}
+}
+
+
+void SQLite3::enableBusyHandler(void)
+{
+	SQLOK(sqlite3_busy_handler(m_database, SQLite3::busyHandler,
+		&m_busy_handler_data), "SQLite3: Failed to set busy handler");
+}
+
+
+// ====================== SQLite3 WAL-sync class implementation ========================
+
+
+bool SQLite3CheckpointThread::shutdown(int timeout)
+{
+	m_shutdown_requested.post();
+	return m_shutdown_acknowledged.wait(timeout * 1000);
 }
 
 
-Database_SQLite3::Database_SQLite3(const std::string &savedir) :
+void *SQLite3CheckpointThread::run()
+{
+	m_db.openDatabase();
+	m_db.setSynchronousLevel();
+
+	infostream << "SQLITE3 WAL sync thread active" << std::endl;
+
+	int checkpoint_interval = WAL_CHECKPOINT_CHECK_INTERVAL;
+
+	bool shutdown = false;
+	do {
+		m_db.checkpointWALPassive();
+		int elapsed = 0;
+		while (!shutdown && elapsed < checkpoint_interval) {
+			checkpoint_interval = g_settings->getU16("sqlite_wal_checkpoint_interval");
```

Comment:
This should be moved out of the loop.  Fetching settings is fairly slow, and you shouldn't do it more than once if you don't have to.